### PR TITLE
Add Upgrade testing workflow

### LINKED ISSUE #216: Automate the DSP Upgrade testing process
Create a GH action to automate the DSP Upgrade testing process. Refer to `Automating the upgrade testing process` section in this ADR for more details. [1]

Focus of this ticket is to set up a test suite for upgrade process, don't need to add a robust set of tests for testing successfull upgrade, that can be a follow up ticket.

[1] https://github.com/opendatahub-io/architecture-decision-records/pull/14

### PR DESCRIPTION
**Pull request by DharmitD**
## The issue resolved by this Pull Request:
Resolves #216 

## Description of your changes:

- This GH workflow sets up a kind cluster, installs OLM, deploys the OpenDataHub operator, deploys DSPO using a KfDef pointing to the most recently released branch specified via the GH action input. It further spins up a DSPA instance.
- The action upgrades DSPO to point to a candidate version, specified in the action input, and spins up a DSPA instance in a separate namespace.
- For now, we've set the most recently released branch to be an input just to be able to run and test this workflow out once merged, but we'll eventually set it to be the `main` branch of `opendatahub/data-science-pipelines-operator` repository in the KfDef.
- Specific upgrade testing checks to ensure that the upgrade went smoothly will be covered in a separate PR.

## Testing instructions

Find a successful workflow run on my fork of the DSPO respository with this same set of configs here, tested out the workflow run multiple times to ensure there aren't any flaky successes: 
https://github.com/DharmitD/data-science-pipelines-operator/actions/workflows/upgrade-test.yml 

## Checklist
- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: .github/actions/kind/action.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        prefix = "$REGISTRY_ADDRESS"
+        insecure = true
+        location = "$REGISTRY_ADDRESS"
+        EOF'
+
+    - name: Setup KinD cluster
+      uses: helm/kind-action@v1.8.0
+      with:
+        cluster_name: cluster
+        version: v0.17.0
```
Comment (zdtsw) [inline .github/actions/kind/action.yml:41]: why not use default 0.20.0?

Comment (DharmitD) [reply .github/actions/kind/action.yml:41]: tried a run with the version set ton 0.20.0, the workflow run failed to set up a kind cluster: https://github.com/DharmitD/data-science-pipelines-operator/actions/runs/6112517009/job/16590146336

Sticking to 0.17.0 as it works for sure.

File: .github/workflows/upgrade-test.yml [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+      - name: Set Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: v1.19
+
+      - name: Set up gotestfmt
+        uses: gotesttools/gotestfmt-action@v2
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (sutaakar) [inline .github/workflows/upgrade-test.yml:41]: Not sure if you need this, do you plan to run go lang tests in here?

File: .github/workflows/upgrade-test.yml [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    timeout-minutes: 60
+    env:
+      OLM_VERSION: v0.24.0
+      VERSION: "v0.0.0-ghaction"  # Need to supply some semver version for bundle to be properly generated
+      CATALOG_BASE_IMG: "registry.access.redhat.com/redhat/community-operator-index:v4.13"
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0  # fetching also previous commits to get tags
```
Comment (sutaakar) [inline .github/workflows/upgrade-test.yml:31]: I guess you don't need this.
I used the complete fetch to be able to retrieve released version based on tag in GitHub repo.

File: .github/workflows/upgrade-test.yml (hunk 2 above)
Comment (DharmitD) [reply .github/workflows/upgrade-test.yml]: My plan was to run some go tests earlier but writing a basic bash script instead now. Done, took this step out.

File: .github/workflows/upgrade-test.yml (hunk 3 above)
Comment (DharmitD) [reply .github/workflows/upgrade-test.yml:31]: I believe this step would be useful for us, in order to retrieve all history of the repo, keeping it for now.

File: .github/resources-crds/buildconfig-crd.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+apiVersion: apiextensions.k8s.io/v1
```
Comment (HumairAK) [inline .github/resources-crds/buildconfig-crd.yaml:1]: can you create a new sub folder and put the resources there, organize it like thus: 

```
.github/resources/
├── crds
│   ├── buildconfig-crd.yaml
│   ├── deploymentconfig-crd.yaml
│   ├── imagestream-crd.yaml
│   └── servicemonitor-crd.yaml
├── kfdef
│   └── kfdef.yaml
├── kind
│   └── kind.yaml
└── olm
    ├── catalogsource.yaml
    ├── operatorgroup.yaml
    └── subscription.yaml

```

Comment (HumairAK) [reply .github/resources-crds/buildconfig-crd.yaml:1]: also in the workflow refer to the `.github/resources` via an env var, and don't hard code it, so something like `.github/resources-kfdef/` should become `${{ env.resources }}/kfdef/`

File: .github/workflows/upgrade-test.yml [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+      - name: Apply CRDs
+        run: |
+          kubectl apply -f .github/resources-crds
+
+      - name: Deploy latest released OpenDataHub operator from OLM
+        id: deploy
+        run: |
+          echo Deploying  operator using Subscription
+          envsubst < .github/resources-olm-install/catalogsource.yaml > ${{ env.TEMP_DIR }}/catalogsource.yaml
```
Comment (HumairAK) [inline .github/workflows/upgrade-test.yml:61]: I am not seeing `TEMP_DIR` being defined anywhere for this job, where is it coming from? is this a gh action internal? I don't see it mentioned here https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables 

It seems like one of the steps defines it as well [here](https://github.com/opendatahub-io/data-science-pipelines-operator/pull/296/files#diff-0c5ad036b2b8cd2db3ed561cf14418fe4da667be18cafdc3167605ed3de8f44dR10), but that would be scoped to the step. Is this being properly substituted?

File: .github/workflows/upgrade-test.yml [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        default: 'v1.x.x'
+        description: 'Candidate version for upgrade'
+        required: true
+
+concurrency:
+  group: ${{ github.head_ref }}-${{ github.workflow }}
+  cancel-in-progress: true
+
+jobs:
+  kubernetes-olm-upgrade:
```
Comment (HumairAK) [inline .github/workflows/upgrade-test.yml:20]:
```suggestion
  dspo-upgrade:
```

File: .github/workflows/upgrade-test.yml [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        env:
+          DSPO_NS: data-science-pipelines-operator
+
+      - name: Deploy a DSPA instance for the candidate DSP Version
+        run: |
+          kubectl create namespace ${{ env.DSPA_NS }}
+          cd ${{ github.workspace }}/config/samples
+          kustomize build . | kubectl -n ${{ env.DSPA_NS }} apply -f -
+        env:
+          DSPA_NS: test-ds-project-2  # Adjust the namespace as needed
```
Comment (HumairAK) [inline .github/workflows/upgrade-test.yml:147]:
```suggestion
          DSPA_NS: test-ds-project-2
```

File: .github/workflows/upgrade-test.yml [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+          DSPO_NS: data-science-pipelines-operator
+
+      - name: Deploy first DSP instance
+        run: |
+          kubectl create namespace ${{ env.DSPA_NS }}
+          cd ${{ env.WORKING_DIR }}/config/samples
+          kustomize build . | kubectl -n ${{ env.DSPA_NS }} apply -f -
+        env:
+          DSPA_NS: test-ds-project-1
+          WORKING_DIR: ${{ github.workspace }}
```
Comment (HumairAK) [inline .github/workflows/upgrade-test.yml:124]: Create a directory from the repo root like so: 

```
tests/upgrades/
└── main.sh
```

Replace this step `Deploy first DSP instance` with a `Run pre-upgrade tests`, in `main.sh` add your deploy dspo code: 

```
kubectl create namespace ${DSPA_NS}
cd ${WORKING_DIR}/config/samples
kustomize build . | kubectl -n ${DSPA_NS} apply -f -
```

You should be able to exclude the `env` in `${{ env.DSPA_NS }}`

Call this script in this step. Follow up task can flesh out the scripts called here for preliminary upgrade tests.

File: .github/workflows/upgrade-test.yml [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+      - name: Print KfDef Manifest Contents
+        run: cat ${{ github.workspace }}/.github/resources-kfdef/kfdef.yaml
+
+      - name: Deploy KfDef Core for the candidate DSP Version
+        run: |
+          kubectl apply -f .github/resources-kfdef/kfdef.yaml -n ${{ env.DSPO_NS }}
+        env:
+          DSPO_NS: data-science-pipelines-operator
+
+      - name: Deploy a DSPA instance for the candidate DSP Version
```
Comment (HumairAK) [inline .github/workflows/upgrade-test.yml:141]: Do the same here replace the step and call the new step `Run upgrade tests`.

**Review by HumairAK — CHANGES_REQUESTED** [review]

File: .github/workflows/upgrade-test.yml (hunk 6 above)
Comment (DharmitD) [reply .github/workflows/upgrade-test.yml]: Done, updated accordingly.

File: .github/workflows/upgrade-test.yml (hunk 7 above)
Comment (DharmitD) [reply .github/workflows/upgrade-test.yml]: Done, updated accordingly.

File: .github/workflows/upgrade-test.yml (hunk 8 above)
Comment (DharmitD) [reply .github/workflows/upgrade-test.yml]: Done, updated to have a separate script and called it here.

File: .github/workflows/upgrade-test.yml (hunk 9 above)
Comment (DharmitD) [reply .github/workflows/upgrade-test.yml]: Done, updated to have a separate script and called it here.

File: .github/resources-crds/buildconfig-crd.yaml (hunk 4 above)
Comment (DharmitD) [reply .github/resources-crds/buildconfig-crd.yaml]: Done, updated to have this directory structure, and refer to the .github/resources via an env var.

File: .github/workflows/upgrade-test.yml [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+      - name: Wait for DSPO deployment to be ready
+        run: |
+          kubectl get deployments -n ${{ env.DSPO_NS }}
+          timeout 300 bash -c 'until [[ $(kubectl get deployment/data-science-pipelines-operator-controller-manager -n '${{ env.DSPO_NS }}') ]]; do sleep 5 && echo "$(kubectl get deployment/data-science-pipelines-operator-controller-manager -n '${{ env.DSPO_NS }}')"; done'
+        env:
+          DSPO_NS: data-science-pipelines-operator
+
+      - name: Run pre-upgrade tests
+        run: |
+          chmod +x tests/upgrades/main.sh
```
Comment (HumairAK) [inline .github/workflows/upgrade-test.yml:118]: you don't need to run this here, just commit the file with these file permissions

File: .github/workflows/upgrade-test.yml [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+      - name: Deploy KfDef Core for the candidate DSP Version
+        run: |
+          kubectl apply -f ${{ env.RESOURCES_DIR }}/kfdef/kfdef.yaml -n ${{ env.DSPO_NS }}
+        env:
+          DSPO_NS: data-science-pipelines-operator
+
+      - name: Run upgrade tests
+        run: |
+          chmod +x tests/upgrades/main.sh
```
Comment (HumairAK) [inline .github/workflows/upgrade-test.yml:141]: same as before

File: .github/workflows/upgrade-test.yml (hunk 5 above)
Comment (DharmitD) [reply .github/workflows/upgrade-test.yml]: The `TEMP_DIR` is being defined right where you've pointed out, it gets stored as an environment variable [here](https://github.com/opendatahub-io/data-science-pipelines-operator/pull/296/files#diff-0c5ad036b2b8cd2db3ed561cf14418fe4da667be18cafdc3167605ed3de8f44dR12), so the scope is GH action wide I believe. @sutaakar could you confirm please?

You can find the `TEMP_DIR` path being correctly substituted in this run on my fork: https://github.com/DharmitD/data-science-pipelines-operator/actions/runs/6113169351/job/16592134955

File: .github/workflows/upgrade-test.yml (hunk 10 above)
Comment (DharmitD) [reply .github/workflows/upgrade-test.yml]: Done, updated the commit with these file permissions and removed it from here.

File: .github/workflows/upgrade-test.yml (hunk 11 above)
Comment (DharmitD) [reply .github/workflows/upgrade-test.yml]: Done, updated the commit with these file permissions and removed it from here.

File: .github/workflows/upgrade-test.yml (hunk 5 above)
Comment (HumairAK) [reply .github/workflows/upgrade-test.yml]: ah yeah I missed the part where they pass it to `>> $GITHUB_ENV`, sounds good.

**Comment by DharmitD** [comment]
Latest run on my fork with these changes: https://github.com/DharmitD/data-science-pipelines-operator/actions/runs/6113169351/job/16592134955

**Pull request merged**